File: flut/flutter/animation/animation_style.py
from typing import Optional, override

from flut._flut.engine import FlutValueObject, _flut_pack_value
from flut._flut.runtime import _flut_unpack_optional_field
from flut.dart.core import Duration
from .curves import Curve


class _AnimationStyle(FlutValueObject):
    _flut_type = "AnimationStyle"

    def __init__(
        self,
        *,
        curve: Optional[Curve] = None,
        duration: Optional[Duration] = None,
        reverseCurve: Optional[Curve] = None,
        reverseDuration: Optional[Duration] = None,
    ):
        super().__init__()
        self.curve = curve
        self.duration = duration
        self.reverseCurve = reverseCurve
        self.reverseDuration = reverseDuration

    @staticmethod
    def _flut_unpack(data: dict) -> "_AnimationStyle":
        return _AnimationStyle(
            curve=_flut_unpack_optional_field(data, "curve"),
            duration=_flut_unpack_optional_field(data, "duration"),
            reverseCurve=_flut_unpack_optional_field(data, "reverseCurve"),
            reverseDuration=_flut_unpack_optional_field(data, "reverseDuration"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.curve is not None:
            result["curve"] = _flut_pack_value(self.curve)
        if self.duration is not None:
            result["duration"] = _flut_pack_value(self.duration)
        if self.reverseCurve is not None:
            result["reverseCurve"] = _flut_pack_value(self.reverseCurve)
        if self.reverseDuration is not None:
            result["reverseDuration"] = _flut_pack_value(self.reverseDuration)
        return result


class AnimationStyle(_AnimationStyle):
    noAnimation = _AnimationStyle(
        duration=Duration.zero,
        reverseDuration=Duration.zero,
    )

File: flut/.flutter/lib/flutter/animation/animation_style.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/dart/core.dart';
import 'package:flut/flutter/animation/curves.dart';

class FlutAnimationStyle extends FlutValueObject {
  final AnimationStyle style;

  const FlutAnimationStyle(this.style) : super('AnimationStyle');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (style.curve != null) {
      result['curve'] = FlutCurve(style.curve!).flutEncode();
    }
    if (style.duration != null) {
      result['duration'] = FlutDuration(style.duration!).flutEncode();
    }
    if (style.reverseCurve != null) {
      result['reverseCurve'] = FlutCurve(style.reverseCurve!).flutEncode();
    }
    if (style.reverseDuration != null) {
      result['reverseDuration'] = FlutDuration(
        style.reverseDuration!,
      ).flutEncode();
    }
    return result;
  }

  static AnimationStyle? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return AnimationStyle(
      curve: runtime.unpackOptionalField<Curve>(data, 'curve'),
      duration: runtime.unpackOptionalField<Duration>(data, 'duration'),
      reverseCurve: runtime.unpackOptionalField<Curve>(data, 'reverseCurve'),
      reverseDuration: runtime.unpackOptionalField<Duration>(
        data,
        'reverseDuration',
      ),
    );
  }
}
